File: 1 basic/1030. 完美数列(25).cpp
#include<iostream>
#include<cstdio>
#include<algorithm>
#define N 100005
using namespace std;

int a[N];

int main(){
	int n,p,max=1;
	cin>>n>>p;
	for(int i=0;i<n;++i) scanf("%d",a+i);
	sort(a,a+n);
	//i<n-max减小搜索范围 
	for(int i=0;i<n-max;++i){
		//j=i+max加快搜索速度 
		int j=i+max;
		//tmp用int会整形溢出 
		for(double tmp=1.0*a[i]*p; j<n && a[j]<=tmp; ++j);
		max = (max>j-i)?max:j-i;
	}
	printf("%d\n",max);
	
	return 0;
} 

File: 1 basic/py/1030. 完美数列(25).py
N,P = list(  map(int, input().split()) )
a = list(  map(int, input().split()) )
a.sort(key = lambda x: x)
i,j,Max = 0,0,1
while i < N - Max :
    j = i + Max
    tmp = a[i] * P  
    while j < N and a[j] <= tmp: j += 1
    if Max < j - i: Max = j - i
    i += 1
print(Max)